tango/step_graph.py

```diff
@@ -151,28 +185,77 @@ def __iter__(self) -> Iterator[str]:
-        while len(steps_not_run) > 0:
-            step = steps_not_run.pop(0)
-            if len(step.dependencies & steps_run) == len(step.dependencies):
-                steps_run.add(step)
-                result.append(step)
-            else:
-                steps_not_run.append(step)
+        for step_name in cls._get_ordered_steps(dependencies):
+            step_dict[step_name].name = step_name
+            result.append(step_dict[step_name])
         return result
 
+    def find_uncacheable_leaf_steps(self) -> Set[Step]:
+        interior_steps: Set[Step] = set()
+        for _, step in self.parsed_steps.items():
+            for dependency in step.dependencies:
+                interior_steps.add(dependency)
+        uncacheable_leaf_steps = {
+            step for step in set(self.values()) - interior_steps if not step.cache_results
+        }
+        return uncacheable_leaf_steps
+
     @classmethod
-    def from_file(cls, filename: PathOrStr):
+    def from_file(cls, filename: PathOrStr) -> "StepGraph":
         params = Params.from_file(filename)
         for package_name in params.pop("include_package", []):
             import_extra_module(package_name)
-        return StepGraph(params.pop("steps", keep_as_dict=True))
+        return StepGraph.from_params(params.pop("steps", keep_as_dict=True))
+
+    def to_config(self):
+        step_dict = {}
+
+        def _to_config(o: Any):
+            if isinstance(o, (list, tuple, set)):
+                return o.__class__(_to_config(i) for i in o)
+            elif isinstance(o, dict):
+                return {key: _to_config(value) for key, value in o.items()}
+            elif isinstance(o, Step):
+                return {"type": "ref", "ref": o.name}
+            elif o is not None and not isinstance(o, (bool, str, int, float)):
+                raise ValueError(o)
+            return o
+
+        for step_name, step in self.parsed_steps.items():
+            try:
+                step_dict[step_name] = {
+                    key: _to_config(value) for key, value in step.config.items()
+                }
+            except ValueError:  # step.config throws an error.
+                # If the step_graph was not constructed using a config, we attempt to create
+                # the config using the step object.
+                step_dict[step_name] = {
+                    key: _to_config(val) for key, val in step._to_params()["kwargs"].items()
+                }
+                step_dict[step_name]["type"] = step.__module__ + "." + step.__class__.__name__
+
+                # We only add cache_results and format to the config if the values are different from default.
+                if step.cache_results != step.CACHEABLE:
+                    step_dict[step_name]["cache_results"] = step.cache_results
+                if step.format != step.FORMAT:
+                    step_dict[step_name]["step_format"] = _to_config(step.format._to_params())
```

Comment:
Nice catch. Did you run into a problem caused by this, or did you just know? I would have forgotten.

Reply:
But wait, shouldn't this happen in `Step._to_params()`?

Reply:
It occurred to me to add this check because I was using `cache_results=False` to test something else earlier. Yes, I think this can be moved to `Step._to_params()` -- I think I wanted to isolate file changes because this is the only place where this was being used.